File: benchmarks/python/tier1/mlp.py
"""MLP benchmark: Linear -> GELU -> LayerNorm, 3 layers."""

import torch
import torch.nn as nn


class MLP(nn.Module):
    def __init__(self):
        super().__init__()
        self.net = nn.Sequential(
            nn.Linear(256, 512),
            nn.GELU(),
            nn.LayerNorm(512),
            nn.Linear(512, 512),
            nn.GELU(),
            nn.LayerNorm(512),
            nn.Linear(512, 256),
        )

    def forward(self, x):
        return self.net(x)


def run(device, batches_per_epoch=100, batch_size=128, **kwargs):
    import sys
    sys.path.insert(0, str(__import__("pathlib").Path(__file__).resolve().parent.parent))
    from harness import run_benchmark

    model = MLP().to(device)
    optimizer = torch.optim.Adam(model.parameters(), lr=1e-3)

    batches = [
        (torch.randn(batch_size, 256, device=device),
         torch.randn(batch_size, 256, device=device))
        for _ in range(batches_per_epoch)
    ]

    return run_benchmark("mlp", model, batches, device, optimizer, **kwargs)

File: benchmarks/src/tier1/mlp.rs
//! MLP benchmark: Linear → GELU → LayerNorm, 3 layers.
//!
//! Tests raw matmul + activation throughput — the simplest possible model.

use flodl::*;
use crate::harness::{BenchConfig, BenchResult, run_benchmark};

pub fn run(device: Device) -> Result<BenchResult> {
    let config = BenchConfig {
        name: "mlp".into(),
        batch_size: 128,
        batches_per_epoch: 100,
        ..Default::default()
    };

    let opts = TensorOptions { dtype: DType::Float32, device };

    // Model: 256 → 512 → 512 → 256
    let model = FlowBuilder::from(Linear::on_device(256, 512, device)?)
        .through(GELU)
        .through(LayerNorm::on_device(512, device)?)
        .through(Linear::on_device(512, 512, device)?)
        .through(GELU)
        .through(LayerNorm::on_device(512, device)?)
        .through(Linear::on_device(512, 256, device)?)
        .build()?;

    let params = model.parameters();
    let param_count = params.iter().map(|p| p.variable.numel()).sum::<i64>() as usize;
    let mut optimizer = Adam::new(&params, 1e-3);
    model.train();

    // Pre-generate synthetic data
    let batches: Vec<(Tensor, Tensor)> = (0..config.batches_per_epoch)
        .map(|_| {
            let x = Tensor::randn(&[config.batch_size as i64, 256], opts).unwrap();
            let y = Tensor::randn(&[config.batch_size as i64, 256], opts).unwrap();
            (x, y)
        })
        .collect();

    run_benchmark(&config, param_count, |_epoch, _warmup| {
        let mut total_loss = 0.0;
        for (x, y) in &batches {
            let input = Variable::new(x.clone(), true);
            let target = Variable::new(y.clone(), false);
            let pred = model.forward(&input)?;
            let loss = mse_loss(&pred, &target)?;

            optimizer.zero_grad();
            loss.backward()?;
            optimizer.step()?;

            total_loss += loss.item()?;
        }
        Ok(total_loss / batches.len() as f64)
    })
}
